[AIRFLOW-3871] render Operators template fields recursively

galak75 opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3871
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
  see AIRFLOW-3871 Jira ticket description

### Tests

- [x] My PR adds the following unit tests :
  4 additional unit tests in `test_python_operator.py`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

galak75 committed:
[AIRFLOW-3871] render Operators template fields recursively

bjoernpollex-sc commented on airflow/models/__init__.py:
```diff
@@ -2452,9 +2454,35 @@ def render_template_from_field(self, attr, content, context, jinja_env):
+                    setattr(content,
+                            attribute,
+                            self.render_template(attribute, value, context))
+                except Exception:
```
I would strongly prefer more granular error handling here. Specifically, any exception raised from `render_template` should be propagated out.

galak75 replied:
My intention was a best effort approach: try to render inner attributes as templates, or else keep attributes unchanged.

e.g: when traversing a `UUID` attribute, `render_template` succeeds, but `setattr` raises an exception (`UUID` is immutable)

But you have a good point: if my inner template expression is wrong, I would like the exception to be propagated out, so I can understand the issue.
I could just put the `render_template` call outside try statement?

I will add some test cases right away

bjoernpollex-sc replied:
Yes, that would be acceptable. I'd also consider logging any exception at least as a warning.

bjoernpollex-sc commented on airflow/models/__init__.py:
```diff
@@ -2181,6 +2181,7 @@ def __init__(
         # Private attributes
         self._upstream_task_ids = set()
         self._downstream_task_ids = set()
+        self._rendered_template_object_ids = set()
```
Ideally I'd like for this to not be a member. This is only used once, during template rendering, there's no need for having it here. The template rendering method could pass it along as parameter. This is not a big issue, it just seems cleaner to me to limit data to the scope where it's actually needed.

galak75 replied:
I would rather keep it as a member to avoid changing `BaseOperator.render_template` and `BaseOperator.render_template_from_field` both signatures.

See `_render_template_object_recursively` implementation

bjoernpollex-sc replied:
That's a fair point.

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -644,9 +648,23 @@ def render_template_from_field(self, attr, content, context, jinja_env):
+            result = self._render_nested_template_fields(content, context)
         return result
 
+    def _render_nested_template_fields(self, content, context):
```
Since _rendered_template_object_ids is only relevant to this function can we instead change the sig to:

```suggestion
    def _render_nested_template_fields(self, content, context, seen_oids):
```

galak75 replied:
@ashb : I have some concern about this change; I thought about it, but it implies changing `BaseOperator.render_template` public method signature to propagate theses already `seen_oids` to nested template fields:
``` python
    def render_template(self, attr, content, context):
```
would become:
```python
    def render_template(self, attr, content, context, seen_oids):
```
This is why I preferred storing this set as an instance variable.
If you think this is not an issue, just let me know and I'll do it.

(see https://github.com/apache/airflow/pull/4743#discussion_r265577416)

ashb replied:
I think this diff should work:


```diff
diff --git a/airflow/models/baseoperator.py b/airflow/models/baseoperator.py
index b71f3c4ca1..3e081a5e3d 100644
--- a/airflow/models/baseoperator.py
+++ b/airflow/models/baseoperator.py
@@ -632,7 +633,10 @@ def render_template_from_field(self, attr, content, context, jinja_env):
         Renders a template from a field. If the field is a string, it will
         simply render the string and return the result. If it is a collection or
         nested set of collections, it will traverse the structure and render
-        all elements in it. If the field has another type, it will return it as it is.
+        all elements in it. For any other type, it will recursively render attributes
+        listed in its 'template_fields' attribute (class or instance level attribute)
+        when this 'template_fields' is defined only.
+        Finally returns the rendered field.
         """
         rt = self.render_template
         if isinstance(content, six.string_types):
@@ -644,9 +648,23 @@ def render_template_from_field(self, attr, content, context, jinja_env):
                 k: rt("{}[{}]".format(attr, k), v, context)
                 for k, v in list(content.items())}
         else:
-            result = content
+            result = self._render_nested_template_fields(content, context, seen_oids=set())
         return result
 
+    def _render_nested_template_fields(self, content, context, seen_oids):
+        if id(content) not in seen_oids:
+            seen_oids.add(id(content))
+            try:
+                nested_template_fields = content.template_fields
+            except AttributeError:
+                # content has no inner template fields
+                return content
+
+            for field in nested_template_fields:
+                rendered = self.render_template(field, getattr(content, field), context, seen_oids=seen_oids)
+                setattr(content, field, rendered)
+        return content
+
     def render_template(self, attr, content, context):
         """
         Renders a template either from a file or directly in a field, and returns
```

But I may have missed something.

galak75 replied:
Hi @ashb,

I might be wrong, but I think there is one change missing in your diff: `render_template` signature has to be changed:

```diff
            for field in nested_template_fields:
                rendered = self.render_template(field, getattr(content, field), context, seen_oids=seen_oids)
                setattr(content, field, rendered)
        return content

-     def render_template(self, attr, content, context):
+     def render_template(self, attr, content, context, seen_oids):
         """
         Renders a template either from a file or directly in a field, and returns
```

And this is precisily the change I'm not sure it 's appropriate. What do you think ?

ashb replied:
Oh yeah, I did miss something.
```diff
            for field in nested_template_fields:
                rendered = self.render_template(field, getattr(content, field), context, seen_oids=seen_oids)
                setattr(content, field, rendered)
        return content

-     def render_template(self, attr, content, context):
+     def render_template(self, attr, content, context, seen_oids=None):
         """
         Renders a template either from a file or directly in a field, and returns
```

And then something like this.

```diff
                 k: rt("{}[{}]".format(attr, k), v, context)
                 for k, v in list(content.items())}
         else:
-            result = content
+            if seen_oids is None:
+                seen_oids = set()
+            result = self._render_nested_template_fields(content, context, seen_oids=seen_oids)
         return result
```

galak75 replied:
Thank you for your feedback. 
I understand you would rather change `BaseOperator.render_template` signature. I will rework this as soon as I can find some time for this (hopefully soon enough!).

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -644,9 +648,23 @@ def render_template_from_field(self, attr, content, context, jinja_env):
                 for k, v in list(content.items())}
         else:
-            result = content
+            result = self._render_nested_template_fields(content, context)
```
```suggestion
            result = self._render_nested_template_fields(content, context, set())
```

(see next comment)

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -644,9 +648,23 @@ def render_template_from_field(self, attr, content, context, jinja_env):
 
+    def _render_nested_template_fields(self, content, context):
+        if id(content) not in self._rendered_template_object_ids:
+            self._rendered_template_object_ids.add(id(content))
```
```suggestion
            seen_oids.add(id(content))
```

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -644,9 +648,23 @@ def render_template_from_field(self, attr, content, context, jinja_env):
         return result
 
+    def _render_nested_template_fields(self, content, context):
+        if id(content) not in self._rendered_template_object_ids:
```
```suggestion
        if id(content) not in seen_oids:
```

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -644,9 +648,23 @@ def render_template_from_field(self, attr, content, context, jinja_env):
+                return content
+
+            for field in nested_template_fields:
+                rendered = self.render_template(field, getattr(content, field), context)
```
```suggestion
                rendered = self.render_template(field, getattr(content, field), context, seen_oids)
```

ashb commented on tests/operators/test_python_operator.py:
```diff
@@ -85,6 +87,10 @@ def setUp(self):
         self.addCleanup(self.dag.clear)
         self.clear_run()
         self.addCleanup(self.clear_run)
+        try:
```
Minor point, but since we are making other changes: could you move this to tearDown as it is cleanup code

galak75 replied:
done

ashb reviewed (changes requested):
Minor changes to the code.

Sorry it sat un-revewied for so long @galak75 

This needs adding to the docs/ tree somewhere - probably in https://airflow.apache.org/concepts.html#id1

I missed your first ping, but please `@ashb` me once you've made these changes. (Oh, though I'm not around for most of May.)

galak75 committed:
[AIRFLOW-3871] move test cleanup code to tearDown method

galak75 committed:
[AIRFLOW-3871] add some documentation about nested template fields

galak75 committed:
[AIRFLOW-3871] replace instance attribute by a method parameter

ashb commented on tests/operators/test_python_operator.py:
```diff
@@ -242,6 +249,510 @@ def test_echo_env_variables(self):
+            }
+        )
+
+    def test_nested_template_fields_can_be_defined_at_the_instance_level(self):
```
I don't think we need this test. I understand what you've done, but this is more testing python classes.

Basically I feel that this test case is just testing this: 

```python
class X:
   x = 1

i = X()
assert i.x == 1
```

and 

```python
class X:
    pass

i = X()
i.x = 1
assert i.x == 1
```

galak75 replied:
:+1:

ashb commented on tests/operators/test_python_operator.py:
```diff
@@ -242,6 +249,510 @@ def test_echo_env_variables(self):
+            start_date=DEFAULT_DATE,
+            state=State.RUNNING
+        )
+        task.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE)
```
Once the template refactor PR is merged we should change all of these to not run the task, but to call the `render_template_fields` directly instead.

galak75 replied:
:+1:

ashb commented on tests/operators/test_python_operator.py:
```diff
@@ -242,6 +249,510 @@ def test_echo_env_variables(self):
+            }
+        )
+
+    def test_inner_attributes_are_rendered_when_marked_as_template_fields_only(self):
```
This looks like it covers the same cases as `test_multiple_template_fields_rendering`?

galak75 replied:
Right. `test_multiple_template_fields_rendering` covers almost every test case...
Anyways, I should use the technique used in the new `TestBaseOperator` class (and move all new test cases in there)

ashb commented on tests/operators/test_python_operator.py:
```diff
@@ -242,6 +249,510 @@ def test_echo_env_variables(self):
                            )
         t.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE)
 
+    def test_template_field_value_rendering(self):
```
We probably don't need this test case.

galak75 replied:
Right. already covered by other tests

ashb reviewed (changes requested):
The docs as you've described them looks sensible, and addresses my previous worries.

A few changes needed in the tests, possibly as part of the rebase after #5461 is merged (which I'd like to get in first)

galak75 committed:
[AIRFLOW-3871] Address PR comments

[AIRFLOW-3871] test nested template fields and fix their rendering

[AIRFLOW-3871] test deep nested template fields

[AIRFLOW-3871] test template rendering on circular references

[AIRFLOW-3871] remove unused attribute

[AIRFLOW-3871] add docstring

[AIRFLOW-3871] remove tests now covered in TestBaseOperator class

[AIRFLOW-3871] move test case on missing nested template field

[AIRFLOW-3871] move test case on jinja invalid expression

[AIRFLOW-3871] cleanup PythonOperator tests

[AIRFLOW-3871] add test case with None value on nested template field

[AIRFLOW-3871] add test cases with None value and empty collections

[AIRFLOW-3871] refactor nested field rendering

[AIRFLOW-3871] extract function to actually render template fields

[AIRFLOW-3871] reuse extracted function for top fields rendering

[AIRFLOW-3871] format code

[AIRFLOW-3871] add typing definition to private method

[AIRFLOW-3871] nested template field rendering now returns nothing

[AIRFLOW-3871] nested template field rendering now returns nothing

galak75 committed:
[AIRFLOW-3871] remove unnecessary 'Optional' parameter marker

ashb reviewed (approved):
Looks good @galak75! 

It's conflicting with master now though :( Can you ping me when you get a chance to resolve it?

galak75 committed:
[AIRFLOW-3871] Fix typo in test method name
